lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp

```diff
@@ -0,0 +1,204 @@
+//===------ YieldOnceCheck.cpp - Check usage of yields in accessors  ------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "yield-once-check"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "llvm/ADT/DenseSet.h"
+#include "swift/SIL/BasicBlockUtils.h"
+#include "swift/AST/DiagnosticsSIL.h"
+
+using namespace swift;
+
+namespace {
+
+class YieldOnceCheck : public SILFunctionTransform {
+
+  template<typename...T, typename...U>
+  static InFlightDiagnostic
+  diagnose(ASTContext &Context, SourceLoc loc, Diag<T...> diag, U &&...args) {
+    return Context.Diags.diagnose(loc, diag, std::forward<U>(args)...);
+  }
+
+  /// A state that is associated with basic blocks to record whether a basic
+  /// block appears before a yield or after a yield, or could be in either
+  /// state, which is denoted by 'Conflict'.
+  enum YieldState {
+    BeforeYield,
+    AfterYield,
+    Conflict
+  };
+
+  struct BBState {
+    YieldState yieldState = BeforeYield;
+    // For AfterYield and Conflict states, track the yield instruction.
+    llvm::Optional<SILInstruction *> yieldInst = None;
+    // For Conflict state, track the first instruction that is found to be in
+    // conflict state. This should be the first instruction of a join node in
+    // the CFG.
+    llvm::Optional<SILInstruction *> conflictInst = None;
+  };
+
+  /// Checks whether there is exactly one yield before a return in every path
+  /// in the control-flow graph.
+  /// Diagnostics are not reported for nodes unreachable from the entry, and
+  /// also for nodes that may not reach the exit of the control-flow graph.
+  void diagnoseYieldOnceUsage(SILFunction &fun) {
+    // Do a traversal of the basic blocks starting from the entry node
+    // in a breadth-first fashion. The traversal order is not important for
+    // correctness, but it could change the errors diagnosed when there are
+    // multiple errors. Breadth-first search could diagnose errors along
+    // shorter paths.
+    llvm::DenseMap<SILBasicBlock*, BBState> visitedBBs;
```

Comment:
As an aside. Please run this through git-clang-format. The patch is inconsistent about the location of the '*'.